### ISSUE TITLE
Variables are not tagged as variables

### ISSUE THREAD

**Issue by jeff-hykin** [jeff-hykin/cpp-textmate-grammar#81]
There are very few instances when a variable is actually tagged as a variable. Ideally all of them would be.

This is a very hard issue, and it may be impossible. This issue is here to track the progress of this particular feature.

Progress
* ✅ member-accessed variable are tagged and all of their members are tagged
* 🔵 #71 Have variable declarations are tagged as variables. (Blocked by many other needed features)
* ❓ variables at the end of scope resolutions `::`
* ❓variables after a class/struct/union/enum
* ❓variables inside of expressions (99% of the cases)

Impossible things for TextMate:
* it is ambiguous to know whether a piece of code is an argumentless macro-call or a variable

**Comment by matter123** [comment]
Determining what B is in A::B is a classic problem that only compiling can solve.

**Comment by matter123** [comment]
function calls and constructor declaration/definitions should be disambiguous.

function calls can only be present where expressions are expected.
Importantly you cannot declare/define a function in an expression.

So determining if `A(B,C)` is a function is dependent on correctly determining where expressions are.

**Comment by jeff-hykin** [comment]
There might be an issue with the following:
`thing1 < thing2 > thing3; `
I think it could be an expression with two comparisons, or it could be a definition of a templates variable. I'd this is true, I think this would also make the vector<vector<>> issue provably unsolvable.

We would need to know if thing1 was a type or a variable, and we just wouldn't have that information.

**Comment by matter123** [comment]
Yeah assuming its unlikely that a comparison is not unused makes sense. Don't forget about on usual uses though. `decltype(a < b > c)` does use (the type of) the return value of the comparison.

**Comment by matter123** [comment]
So I created the following code snippet
```c++
#include<iostream>
template<int i=10>
constexpr int thing1 = 1*2;
constexpr int thing2 = 20;
constexpr int thing3 = 0;
int main(){
	 thing1<thing2>thing3;
	 std::cout << thing3;
}
```
As far as I can tell it should be ambiguous (with c++17) what `thing1<thing2>thing3` represents. The compilers seem to treat it as two unused comparisons.

**Comment by jeff-hykin** [comment]
That ambiguity is really annoying. And good thinking about decltype.

The good news is, if it's inside of decltype we can probably assume it's an expression. If it was a type, they wouldn't need it to be decltype.

Another edge case to consider is C-style type castings. I'm not sure if they're non-ambiguous.

**Comment by matter123** [comment]
Detecting a c style cast might be annoying but the cast expression always is an expression

**Comment by matter123** [comment]
And I'm mistake it turns out c++17, doesn't have variable type argument deduction so that is declaring a new variable thing3.

**Comment by matter123** [comment]
And I'm mistake it turns out c++17, doesn't have variable type argument deduction so that is declaring a new variable thing3.

I'll make a new example using ctad and implicit conversion maybe.

**Comment by matter123** [comment]
Well, I can't find an example that is truly ambiguous from the compiler's perspective, but the compiler seems to prefer a couple of comparisons. over a template

**Comment by matter123** [comment]
regarding variables after struct/class

In C the following pattern is fairly common

```c
typedef struct point {
    int x, y;
} point;
```
this allows writing `double distance(point p1, point p2);`

Detecting variables after a struct/class require keeping the knowledge of whether or not the declaration was preceded by `typedef`.

I guess a range that start with typedef and includes a special class/struct pattern?

**Comment by jeff-hykin** [comment]
I think that range will work.

I noticed too (we should probably make an issue about it) that in C it seems common to use syntax like
`void func(struct aStructName value, int  a);`
or
`struct aStructName var;`

I made some special-case patterns so that structs are pretty much covered but I think the same style declarations can be done with enums and unions.

**Comment by matter123** [comment]
There is already an issue for types in C preceded by storage class specifiers #46. I did expand upon it slightly.
